[DirectX] Infrastructure to collect shader flags for each function

bharadwajy opened the pull request:
Currently, ShaderFlagsAnalysis pass represents various module-level properties as well as function-level properties of a DXIL Module using a single mask. However, one mask per function is needed for accurate computation of shader flags mask, such as for entry function metadata creation.

This change introduces a structure that wraps a sorted vector of function-shader flag mask pairs that represent function properties instead of a single shader flag mask that represents module properties and properties of all functions. The result type of ShaderFlagsAnalysis pass is changed to newly-defined structure type instead of a single shader flags mask. 

This allows accurate computation of shader flags of an entry function (and all functions in a library shader) for use during its metadata generation (DXILTranslateMetadata pass) and its feature flags in DX container globals construction (DXContainerGlobals pass) based on the shader flags mask of functions. However, note that the change to implement propagation of such callee-based shader flags mask computation is planned in a follow-on PR. Consequently, this PR changes shader flag mask computation in DXILTranslateMetadata and DXContainerGlobals passes to simply be a union of module flags and shader flags of all functions, thereby retaining the existing effect of using a single shader flag mask.

Commit:
[NFC][DirectX] Infrastructure to collect shader flags for each function

Currently, ShaderFlagsAnalysis pass represents various module-level
properties as well as function-level properties of a DXIL Module
using a single mask. However, separate flags to represent module-level
properties and function-level properties are needed for accurate computation
of shader flags mask, such as for entry function metadata creation.

This change introduces a structure that allows separate representation of

(a) shader flag mask to represent module properties
(b) a map of function to shader flag mask that represent function properties

instead of a single shader flag mask that represents module properties
and properties of all function. The result type of ShaderFlagsAnalysis
pass is changed to newly-defined structure type instead of a single shader
flags mask.

This seperation allows accurate computation of shader flags of an entry
function for use during its metadata generation (DXILTranslateMetadata pass)
and its feature flags in DX container globals construction (DXContainerGlobals
pass) based on the shader flags mask of functions called in entry function.
However, note that the change to implement such callee-based shader flags mask
computation is planned in a follow-on PR. Consequently, this PR changes shader
flag mask computation in DXILTranslateMetadata and DXContainerGlobals passes
to simply be a union of module flags and shader flags of all functions, thereby
retaining the existing effect of using a single shader flag mask.

Commit:
clang-format changes

bogner commented:
We should be able to write tests that the shader flags are correct per-function by running the analysis on some IR and printing the results, which we can then check.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -65,15 +73,25 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+  OS << "; Shader Flags mask for Module:\n";
+  Flags.ModuleFlags.print(OS);
+  for (auto SF : Flags.FuncShaderFlagsMap) {
+    OS << "; Shader Flags mash for Function: " << SF.first->getName() << "\n";
```
```suggestion
    OS << "; Shader Flags mask for Function: " << SF.first->getName() << "\n";
```

bharadwajy replied:
Fixed.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -65,15 +73,25 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
-  ComputedShaderFlags Flags = AM.getResult<ShaderFlagsAnalysis>(M);
-  Flags.print(OS);
+  DXILModuleShaderFlagsInfo Flags = AM.getResult<ShaderFlagsAnalysis>(M);
+  OS << "; Shader Flags mask for Module:\n";
```
This looks like it is actually a functional change?  In which case we should have some tests to test the new functionality that's been added.

bharadwajy replied:
> This looks like it is actually a functional change? In which case we should have some tests to test the new functionality that's been added.

Made changes in tests.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -20,29 +20,37 @@
 
-static void updateFlags(ComputedShaderFlags &Flags, const Instruction &I) {
+static void updateFlags(DXILModuleShaderFlagsInfo &MSFI, const Instruction &I) {
+  ComputedShaderFlags &FSF = MSFI.FuncShaderFlagsMap[I.getFunction()];
```
It looks like this is being called inside a loop that looks something like:

```
for (const auto& BB: F) 
  for (const auto &I : BB)
    update(MSGI, I);
```

So say we have 100 functions with 1,000 instructions in each one, this ends up looking the function up in the map 100,000 times. If it were rearranged so that the lookup happens once per function this could be reduced to just 100 times.

Something like:

```
for (const auto& BB: F) {
  ComputedShaderFlags &CSF = MSFI.FuncShaderFlagsMap[F];
  for (const auto &I : BB) 
    update(CSF, I);
}

bharadwajy replied:
Changed.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -20,29 +20,37 @@
 
-static void updateFlags(ComputedShaderFlags &Flags, const Instruction &I) {
+static void updateFlags(DXILModuleShaderFlagsInfo &MSFI, const Instruction &I) {
+  ComputedShaderFlags &FSF = MSFI.FuncShaderFlagsMap[I.getFunction()];
```
FSF -> CSF?

bharadwajy replied:
> FSF -> CSF?

Changed.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -20,29 +20,37 @@
+  for (const auto &F : M) {
+    if (F.isDeclaration())
+      continue;
+    if (!MSFI.FuncShaderFlagsMap.contains(&F)) {
```
Can F appear multiple times in M?  Or in multiple modules?  I'm wondering why this is needed here.

Do we have test coverage for this case?

bharadwajy replied:
> Can F appear multiple times in M? Or in multiple modules? I'm wondering why this is needed here.
> 
> Do we have test coverage for this case?

F can appear as a declaration and a definition in a module. The former is ignored. This check is modified in the latest changeset to fail for any ill-formed modules.

Commit:
Use a SmallVector of pairs instead of DenseMap to collect Function
pointers and corresponding shader flag masks. This follows the
recommendations in LLVM Programmer's Manual as the current usage
pattern has distinct phases of insertion of computed shader flags
followed by querying. Upon insertion, the Smallvector is sorted and
binary search is used for querying. Necessary comparison function of
pairs is also implemented.

Added a simple DiagnosticInfoShaderFlags for emitting diagnostics.

Added tests to verify shader flags masks collected per-function.

bharadwajy commented:
> We should be able to write tests that the shader flags are correct per-function by running the analysis on some IR and printing the results, which we can then check.

Added per-function test.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,115 @@
+    // mask of the function should be present.
+    if (MSFI.hasShaderFlagsMask(&F)) {
+      M.getContext().diagnose(DiagnosticInfoShaderFlags(
+          M, "Shader Flags mask for Function '" + Twine(F.getName()) +
```
> M, "Shader Flags mask for Function '" + Twine(F.getName()) +

Is the `Twine` really necessary here?

damyanp replied:
If the `Twine` isn't needed here, is it needed in any of the other places in this file where it's just wrapped around a call to `getName()`?

bharadwajy replied:
> If the `Twine` isn't needed here, is it needed in any of the other places in this file where it's just wrapped around a call to `getName()`?

Changed return type to `Expected<T>` and deleted `class DiagnosticInfoShaderFlags`.

damyanp reviewed (changes requested):
I stopped reading this after seeing the string construction in the comparison function.  I'm either missing something big about this change, which means I'd need to review again once I've been corrected, or this PR is going to be somewhat different anyway, so I'll wait until that's resolved before looking at it again.

bharadwajy commented:
> I stopped reading this after seeing the string construction in the comparison function. I'm either missing something big about this change, which means I'd need to review again once I've been corrected, or this PR is going to be somewhat different anyway, so I'll wait until that's resolved before looking at it again.

Lexicographic sort of functions by their signatures seemed the appropriate. I looked at using [`FunctionComparator::compareSignature()`](https://github.com/llvm/llvm-project/blob/7a710110fcb2ad5d903ec41ba6a63193cf03edc0/llvm/lib/Transforms/Utils/FunctionComparator.cpp#L946) but did not see its usage in the sources other than in tests, so wasn't sure if its usage is encouraged or not.

What are the concerns?

damyanp commented:
> What are the concerns?

Have a think about how many strings will be built, and how many memory allocations there'll be, every time `sort` or `lower_bound` needs to compare two elements of the vector.

Commit:
Compare functions by their names instead of constructing a pseudo-signature
Non-empty Function names are unique in LLVM IR.

Update the expected test output accordingly

bharadwajy commented:
> > What are the concerns?
> 
> Have a think about how many strings will be built, and how many memory allocations there'll be, every time `sort` or `lower_bound` needs to compare two elements of the vector.

Changed to compare functions by their names instead of constructing a pseudo-signature. Non-empty Function names are unique in LLVM IR.

bharadwajy commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,43 +13,114 @@
+    if (MSFI.hasShaderFlagsMask(&F)) {
+      M.getContext().diagnose(DiagnosticInfoShaderFlags(
+          M, "Shader Flags mask for Function '" + Twine(F.getName()) +
+                 "' already exits"));
```
Fixed.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,88 @@
+      continue;
+    // Each of the functions in a module are unique. Hence no prior shader flags
+    // mask of the function should be present.
+    if (MSFI.hasShaderFlagsMask(&F)) {
```
`hasShaderFlagsMask` does a binary search through `FuncShaderFlagsVec`.  However, `FuncShaderFlagsVec` is not necessarily sorted, because new entries are added inside the loop and it is only sorted after the loop exits.

This is probably moot since the functions in the module are meant to be unique, which then makes me wonder why this check is here at all.

Commit:
Fix incorrect lookup that expects a sorted vector done in a possibly unsorted vector
Other changes based on latest PR feedback

Commit:
Use getFunctionList() instead of iterating Module for functions.

Commit:
Change return type of getShaderFlagsMask() to Expected<T>
Delete unused class DiagnosticInfoShaderFlags

bharadwajy commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,16 +115,53 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+  if (Iter == FuncShaderFlagsVec.end()) {
+    Func->getContext().diagnose(DiagnosticInfoShaderFlags(
+        *(Func->getParent()), "Shader Flags information of Function '" +
+                                  Twine(Func->getName()) + "' not found"));
```
> > When can this happen? Does it actually indicate a bug in the code or is there some input that someone could generate that would cause this diagnostic to get emitted?
> 
> Since this is the method used by other passes to query for shader flags mask for a function, a failed query is reported as a diagnostic. If the preference is to return a value indicating an error - allowing the caller to determine the consequences - I am open to that as well.

Changed return type to `Expected<T>` and deleted `class DiagnosticInfoShaderFlags`.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -318,22 +313,39 @@ static void translateMetadata(Module &M, const DXILResourceMap &DRM,
+    // Create a consolidated shader flag mask of all functions in the library
+    // to be used as shader flags mask value associated with top-level library
+    // entry metadata.
+    uint64_t ConsolidatedMask = ShaderFlags.ModuleFlags;
+    for (const auto &FunFlags : ShaderFlags.FuncShaderFlagsVec) {
+      ConsolidatedMask |= FunFlags.second;
+    }
```
Do we really want a mask of both the module level and consolidated function flags here? That seems a bit surprising.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -318,22 +313,39 @@ static void translateMetadata(Module &M, const DXILResourceMap &DRM,
-                                                                 : ShaderFlags;
+    Expected<const ComputedShaderFlags &> ECSF =
+        ShaderFlags.getShaderFlagsMask(EntryProp.Entry);
+    if (Error E = ECSF.takeError()) {
+      M.getContext().diagnose(DiagnosticInfoTranslateMD(
+          M, "Shader Flags information of Function '" +
+                 Twine(EntryProp.Entry->getName()) + "' not found"));
+    }
```
If we're going to completely ignore the error here it would make more sense to simply have `getShaderFlagsMask` return `nullptr` if it fails. Using `Expected` to return a string that says "not found" is kind of heavy handed anyways.

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions-obj-test.ll:
```diff
@@ -0,0 +1,19 @@
+; RUN: llc %s --filetype=obj -o - | obj2yaml | FileCheck %s --check-prefix=DXC
```
If we only have one command line it's better to use the default `CHECK` prefix.

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions-obj-test.ll:
```diff
@@ -0,0 +1,19 @@
+; DXC: - Name:            SFI0
+; DXC-NEXT:     Size:            8
+; DXC-NEXT:     Flags:
+; DXC-NEXT:       Doubles:         true
+; DXC-NOT:   {{[A-Za-z]+: +true}}
+; DXC:            DX11_1_DoubleExtensions:         true
+; DXC-NOT:   {{[A-Za-z]+: +true}}
+; DXC:       NextUnusedBit:   false
```
What does this print that we need this awkward CHECK-NOT of anything that's specifically "true" rather than just a CHECK-NEXT?

bogner replied:
This doesn't seem like the place to check that all of the other flags are false. Just doing the two checks should be sufficient:
```
; CHECK: Doubles: true
; CHECK: DX11_1_DoubleExtensions: true
```

bharadwajy replied:
> This doesn't seem like the place to check that all of the other flags are false. Just doing the two checks should be sufficient:
> 
> ```
> ; CHECK: Doubles: true
> ; CHECK: DX11_1_DoubleExtensions: true
> ```

Changes made to limited checking to the two flags.

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,16 +97,41 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
 }
 
 PreservedAnalyses ShaderFlagsAnalysisPrinter::run(Module &M,
                                                   ModuleAnalysisManager &AM) {
-  ComputedShaderFlags Flags = AM.getResult<ShaderFlagsAnalysis>(M);
+  DXILModuleShaderFlagsInfo Flags = AM.getResult<ShaderFlagsAnalysis>(M);
   Flags.print(OS);
   return PreservedAnalyses::all();
 }
```
This should really loop over all of the functions in the module, look each one up in the flags analysis, and print them individually. This way the flags will be printed in the same order as the functions show up in the module, which makes the tests for this a lot more reasonable. I would probably drop the `DXILModuleShaderFlagsInfo::print` method entirely and just do it here.

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,70 @@
-  ComputedShaderFlags Flags;
-  for (const auto &F : M)
-    for (const auto &BB : F)
+static bool compareFunctions(Function const *F1, Function const *F2) {
+  return (F1->getName().compare(F2->getName()) < 0);
+}
+
+static bool compareFuncSFPairs(const FuncShaderFlagsMask &First,
+                               const FuncShaderFlagsMask &Second) {
+  return compareFunctions(First.first, Second.first);
+}
```
Why are we sorting by function name at all here? We don't need the structure to be in a deterministic order if we're just doing lookups on it, so we should really just sort all of these by address.

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,88 @@
-  for (const auto &F : M)
+static bool compareFuncSFPairs(const FuncShaderFlagsMask &First,
+                               const FuncShaderFlagsMask &Second) {
+  return (First.first->getName().compare(Second.first->getName()) < 0);
```
The pattern of building the vector and sorting it is fine, but we really don't need to sort this by name.  Just sorting by the pointer addresses would be better since we don't really need a deterministic order.

Commit:
Address PR feedback
- Use CSF.Doubles directly
- Remove type-aliasies FuncShaderFlags*
- Make ModuleFlags and FunctionFlags private
- Delete DXILModuleShaderFlagsInfo::print()
- Delete check prefix DXC from test with a single run
- Get rid of compare functions
- Change order of expected output accordingly in double-extensions.ll
- Add extra comments for clarification
- Add back DiagnosticInfoShaderFlags
- Additional error checks

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,91 @@
-  for (const auto &F : M)
-    for (const auto &BB : F)
+static DXILModuleShaderFlagsInfo computeFlags(const Module &M) {
+  DXILModuleShaderFlagsInfo MSFI;
```
Suggest moving declaration to before the loop where it's collected.  I don't think it is accessed at all while building up the list of functions.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,91 @@
+  }
+  llvm::sort(FuncList);
+
+  MSFI.clear();
```
Why does it need to be cleared after construction?  Is it not constructed in a "clear" state?

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,91 @@
+  MSFI.clear();
+
+  // Collect shader flags for each of the functions
+  for (const auto &F : FuncList) {
```
```suggestion
  for (const auto *F : FuncList) {
```

or:

```suggestion
  for (const Function *f : FuncList) {
```

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,91 @@
+  // Construct a sorted list of functions in the module. Walk the sorted list to
+  // create a list of <Function, Shader Flags Mask> pairs. This list is thus
+  // sorted at construction time and may be looked up using binary search.
+  SmallVector<Function const *> FuncList;
```
I think most LLVM code would write it like this:

```suggestion
  SmallVector<const Function *> FuncList;
```

Although, if there isn't an existing coding convention here then the way you've written it is, FWIW, my preference :)

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,17 +118,78 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+  FunctionFlags.clear();
+}
+
+/// Insert the pair <Func, FlagMask> into the sorted vector
+/// FunctionFlags. The insertion is expected to be in-order and hence
+/// is done at the end of the already sorted list.
+[[nodiscard]] bool DXILModuleShaderFlagsInfo::insertInorderFunctionFlags(
+    const Function *Func, ComputedShaderFlags FlagMask) {
+  std::pair<Function const *, ComputedShaderFlags> V{Func, {}};
+  auto Iter = llvm::lower_bound(FunctionFlags, V);
+  if (Iter != FunctionFlags.end())
+    return false;
+
+  FunctionFlags.push_back({Func, FlagMask});
+  return true;
+}
```
This all seems overly complicated.  Why isn't it sufficient just to do a push_back from within `computeFlags`?

The only way I think this could go wrong would be if something the call to `llvm::sort(FuncList)` somehow didn't sort the list.  But as some point I think we have to trust that that code works.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,17 +118,78 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+
+SmallVector<std::pair<Function const *, ComputedShaderFlags>>
+DXILModuleShaderFlagsInfo::getFunctionFlags() const {
+  return FunctionFlags;
```
This is returning a copy of `FunctionFlags`.  Is that really the intent here?  I'd expect it to return a const-reference to `FunctionFlags` to avoid unnecessary copies being made.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,17 +118,78 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+Expected<const ComputedShaderFlags &>
+DXILModuleShaderFlagsInfo::getShaderFlagsMask(const Function *Func) const {
+  std::pair<Function const *, ComputedShaderFlags> V{Func, {}};
+  // It is correct to delegate comparison of two pairs, say P1, P2, to default
+  // operator< for pairs that returns the evaluation of (P1.first < P2.first)
+  // viz., comparison of Function pointers - the same comparison criterion used
+  // for sorting module functions walked to form FunctionFLags vector..
```
TBH, this comment leaves me more confused about the code than I'd be without it.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,17 +118,78 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+  // operator< for pairs that returns the evaluation of (P1.first < P2.first)
+  // viz., comparison of Function pointers - the same comparison criterion used
+  // for sorting module functions walked to form FunctionFLags vector..
+  auto Iter = llvm::lower_bound(FunctionFlags, V);
+  if (Iter == FunctionFlags.end()) {
+    return createStringError("Shader Flags information of Function '" +
+                             Func->getName() + "' not found");
+  }
```
This isn't the correct way to use `lower_bound`.  I think you probably meant to use `binary_search` here with a custom `operator<` that compares `Function*`s with `pair<Function*, ComputedShaderFlags>`.

If you do want to use `lower_bound`, then you need to account for it returning either `end()`, or an iterator pointing to the first element that is not before the one you're searching for. So you'd normally see a test like this:

```
auto it = lower_bound(FunctionFlags, V);
if (it == FunctionFlags.end() || it->first != Func) {
   return error;
}
```

As it is now, if you had this in your vector: `A B D` and you searched for `C` then you'd get an iterator pointing at `D` and so you'd return the flags for `D`.

Commit:
Address PR feedback

Commit:
Address PR feedback.
Move the functionality of static void updateFlags(...) to private
method void DXILModuleShaderFlagsInfo::updateFuctionFlags(...) and
that of static DXILModuleShaderFlagsInfo computeFlags(const Module &M)
to public method bool DXILModuleShaderFlagsInfo::initialize(const Module &M).

Commit:
Delete unused function

damyanp commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -318,22 +314,38 @@ static void translateMetadata(Module &M, const DXILResourceMap &DRM,
 
-  if (MMDI.ShaderProfile == Triple::EnvironmentType::Library)
+  if (MMDI.ShaderProfile == Triple::EnvironmentType::Library) {
+    // Create a consolidated shader flag mask of all functions in the library
+    // to be used as shader flags mask value associated with top-level library
+    // entry metadata.
+    uint64_t ConsolidatedMask = ShaderFlags.getModuleFlags();
+    for (const auto &FunFlags : ShaderFlags.getFunctionFlags()) {
+      ConsolidatedMask |= FunFlags.second;
+    }
```
I disagree - but it's up to you whether or not you take this feedback.  `DXILModuleShaderFlagsInfo` is a reasonably good abstraction to be working with, but the moment you expose all of its members via accessors that are then exclusively used to calculate other values you're breaking that abstraction.  If this is the way that the class is intended to be used then you might as well remove all the encapsulation and make the members public, and just use const-ness as a strong hint not to modify the member variables.

bharadwajy replied:
OK. Deleted the access function `getFunctionFlags()`.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -60,21 +61,47 @@ struct ComputedShaderFlags {
+  void insertInorderFunctionFlags(const Function *, ComputedShaderFlags);
+
+private:
+  // Shader Flag mask representing module-level properties. These are
+  // represented using the macro DXIL_MODULE_FLAG
+  ComputedShaderFlags ModuleFlags;
```
So there's no code that uses it currently and nothing testing that it is correctly factored in to any aggregated values? I don't think we should be checking in untested code.

bharadwajy replied:
Deleted the field `ModuleFlags`.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,78 @@
+    FunctionFlags.push_back({&F, CSF});
+  }
+  llvm::sort(FunctionFlags);
+  return true;
```
Why have a return value if it'll always return true?  Are there ever likely to be any failure cases?

bharadwajy replied:
Changed `initialize()` to a constructor.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -60,21 +61,48 @@ struct ComputedShaderFlags {
+  // of the functions in the module. Shader Flags of each function are those
+  // represented using the macro SHADER_FEATURE_FLAG.
+  SmallVector<std::pair<Function const *, ComputedShaderFlags>> FunctionFlags;
+  void updateFuctionFlags(ComputedShaderFlags &CSF, const Instruction &I);
```
Probably best not to say that out loud in polite conversation :)

```suggestion
  void updateFunctionFlags(ComputedShaderFlags &CSF, const Instruction &I);
```

bharadwajy replied:
Fixed the spelling :)

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,17 +105,57 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+}
+
+bool ShaderFlagsAnalysisWrapper::runOnModule(Module &M) {
+  MSFI.initialize(M);
```
We talked a bit about this a bit previously, and it is something that I tend to worry a lot about, but I think we should generally prefer to use a constructor rather than a construct/initialize pattern if there's no way for initialization to fail.

Do we also need to protect against `ShaderFlagsAnalysisWrapper::getShaderFlags` being called before `runOnModule`, at least with an assert?  If so then maybe make MSFI be `std::optional<DXILModuleShaderFlagsInfo>` would be a way to let you write that assert.

bharadwajy replied:
Replaced `initialize()` with a constructor.

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,17 +114,63 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+Expected<const ComputedShaderFlags &>
+DXILModuleShaderFlagsInfo::getShaderFlagsMask(const Function *Func) const {
+  std::pair<Function const *, ComputedShaderFlags> V{Func, {}};
+  const auto *Iter = llvm::lower_bound(FunctionFlags, V);
```
Ah, so clang-tidy thinks that lower_bound returns a pointer to the pair, rather than an explicit iterator type. I'm not 100% sure that clang-tidy isn't being a bit too overzealous here resolving the type aliases.

damyanp reviewed (commented):
Thanks, this is looking good. I think my remaining opens are really nits.

We should look for someone more expert in this area to approve though.

Commit:
Use function shader flag mask for module flags; track combined mask

Delete DXILModuleShaderFlagsInfo::ModuleFlags and track module flags
in shader flags mask of each function.

Add private field DXILModuleShaderFlagsinfo::CombinedSFMask to
represent combined shader flags masks of all functions. Update the
value as it is computed per function.

Change DXILModuleShaderFlagsInfo::initialize(Module&) to constructor

damyanp commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -120,16 +127,17 @@ class ShaderFlagsAnalysisPrinter
 /// passes which run through the legacy pass manager.
 class ShaderFlagsAnalysisWrapper : public ModulePass {
-  DXILModuleShaderFlagsInfo MSFI;
+  std::unique_ptr<DXILModuleShaderFlagsInfo> MSFI;
```
Out of interest, did you consider `std::optional` for this?

bharadwajy replied:
> +1
> 
> I also question if we need the optional. Why was the old pattern of the uninitialized flag structure a problem?

Reverted to old pattern; changed to use `initialize()` instead of a constructor to avoid dynamic memory allocation about which concerns were expressed.

damyanp reviewed (commented):
LGTM, but would look for a review from someone with more domain knowledge than I have.

Commit:
Rename operator |= as merge

llvm-beanz commented on llvm/lib/Target/DirectX/DXContainerGlobals.cpp:
```diff
@@ -78,13 +78,16 @@ bool DXContainerGlobals::runOnModule(Module &M) {
-                                .getFeatureFlags());
+  // TODO: Feature flags mask is obtained as a collection of feature flags
+  // of the shader flags of all functions in the module. Need to verify
+  // and modify the computation of feature flags to be used.
```
Do we have an issue tracking this?

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,79 @@
+/// for Shader Flags Analysis pass
+class DiagnosticInfoShaderFlags : public DiagnosticInfo {
+private:
+  const Twine &Msg;
```
I get really nervous when someone stores a Twine. This effectively means that you must create and destroy this object in a single expression, otherwise the Twine or its attached arguments can go out of scope and you have a memory error.

It seems to me like what you really need is an adapter that converts an `llvm::Error` to a `DiagnosticInfo`, so that you can just pass the Error object right through.

We should add a utility to llvm/Support/Error to facilitate that.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,79 @@
+
+void DXILModuleShaderFlagsInfo::updateFunctionFlags(ComputedShaderFlags &CSF,
+                                                    const Instruction &I) {
+  if (!CSF.Doubles) {
+    CSF.Doubles = I.getType()->isDoubleTy();
+  }
```
nit: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
  if (!CSF.Doubles)
    CSF.Doubles = I.getType()->isDoubleTy();
```

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,79 @@
+    CSF.Doubles = I.getType()->isDoubleTy();
+  }
+  if (!CSF.Doubles) {
+    for (Value *Op : I.operands()) {
+      CSF.Doubles |= Op->getType()->isDoubleTy();
+    }
```
nit: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
    for (Value *Op : I.operands())
      CSF.Doubles |= Op->getType()->isDoubleTy();
```

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,79 @@
     case Instruction::FPToUI:
     case Instruction::FPToSI:
-      Flags.DX11_1_DoubleExtensions = true;
+      // TODO: To be set if I is a call to DXIL intrinsic DXIL::Opcode::Fma
```
Do we have an issue for this?

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,20 +106,64 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
 }
 
+/// Get the combined shader flag mask of all module functions.
+const ComputedShaderFlags DXILModuleShaderFlagsInfo::getCombinedFlags() const {
```
nit: a one-line function like this we should probably declare inline in the header so that it can be inlined at calls.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,20 +106,64 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+DXILModuleShaderFlagsInfo::getShaderFlagsMask(const Function *Func) const {
+  std::pair<Function const *, ComputedShaderFlags> V{Func, {}};
+  const auto Iter = llvm::lower_bound(FunctionFlags, V);
+  if (Iter == FunctionFlags.end() || Iter->first != Func) {
+    return createStringError("Shader Flags information of Function '" +
+                             Func->getName() + "' not found");
+  }
```
nit: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements
```suggestion
  if (Iter == FunctionFlags.end() || Iter->first != Func)
    return createStringError("Shader Flags information of Function '" +
                             Func->getName() + "' not found");
```

bogner replied:
The only way to have this fail is if we've invalidated the analysis (and failed to tell the pass manager) or we're trying to use it wrong. This should just be an assert.

bharadwajy replied:
> nit: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

Changed the `if` statement `assert`.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,20 +106,64 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
                                                   ModuleAnalysisManager &AM) {
-  ComputedShaderFlags Flags = AM.getResult<ShaderFlagsAnalysis>(M);
-  Flags.print(OS);
+  DXILModuleShaderFlagsInfo FlagsInfo = AM.getResult<ShaderFlagsAnalysis>(M);
```
You could store the `const &` to avoid the copy here.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,20 +106,64 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+      continue;
+    OS << "; Shader Flags mask for Function: " << F.getName() << "\n";
+    auto SFMask = FlagsInfo.getShaderFlagsMask(&F);
+    if (Error E = SFMask.takeError()) {
+      M.getContext().diagnose(
+          DiagnosticInfoShaderFlags(M, toString(std::move(E))));
+    }
```
This should be done through a call to `llvm::handleAllErrors`. Something like:

```suggestion
   if (!SFMask)
      return handleAllErrors(std::move(E),
                    [&](std::unique_ptr<ErrorInfoBase> EIB) -> Error {
                      M.getContext().diagnose(errorToDiagnosticInfo(EIB);
                      return Error::success();
                    });
```

This handles arrays of errors so that your function can return more than one error.

bharadwajy replied:
> This should be done through a call to `llvm::handleAllErrors`. Something like:
> 
> This handles arrays of errors so that your function can return more than one error.

Deleted this error-handling code as a result of the assertion added in `getShaderFlagsmask()`.

llvm-beanz commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,79 @@
+  for (const auto &F : M.getFunctionList()) {
+    if (F.isDeclaration())
+      continue;
+    ComputedShaderFlags CSF{};
```
nit: ComputedShaderFlags has a default constructor to zero itself out, the empty initializer list is unnecessary.
```suggestion
    ComputedShaderFlags CSF;
```

llvm-beanz commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -318,23 +314,31 @@ static void translateMetadata(Module &M, const DXILResourceMap &DRM,
-                                                                 : ShaderFlags;
+    Expected<const ComputedShaderFlags &> EntrySFMask =
+        ShaderFlags.getShaderFlagsMask(EntryProp.Entry);
+    if (Error E = EntrySFMask.takeError()) {
+      M.getContext().diagnose(
+          DiagnosticInfoTranslateMD(M, toString(std::move(E))));
+    }
```
This should also be a `handleAllErrors` call.

bharadwajy replied:
> This should also be a `handleAllErrors` call.

Deleted this error-handling code as a result of the assertion added in `getShaderFlagsmask()`.

llvm-beanz reviewed (commented):
Something I'm also just noticing. It doesn't look like any of the tests here actually trigger the errors. Should these even be errors? Should they instead be asserts?

Commit:
Address PR feedback

Delete DiagnosticInfoShaderFlags and flag failure to find shader flags mask
of a function as fatal error as it is not expected to be the case
for well-formed modules.

Restore ModuleFlags::initialize inplace of constructor

Rename DXILModuleShaderFlagsInfo as ModuleShaderFlags

Use getModuleFlags() in operator uint_64()

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -63,20 +106,64 @@ void ComputedShaderFlags::print(raw_ostream &OS) const {
+/// else an error
+Expected<const ComputedShaderFlags &>
+DXILModuleShaderFlagsInfo::getShaderFlagsMask(const Function *Func) const {
+  std::pair<Function const *, ComputedShaderFlags> V{Func, {}};
+  const auto Iter = llvm::lower_bound(FunctionFlags, V);
```
Probably simpler to use a comparator function that just compares against the pair's `first` rather than constructing a sentinel to compare against.

Also llvm style prefers `const Function *` to `Function const *`

bharadwajy replied:
Changed to use comparator.

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -60,21 +62,54 @@ struct ComputedShaderFlags {
 
+struct DXILModuleShaderFlagsInfo {
+  DXILModuleShaderFlagsInfo(const Module &);
+  Expected<const ComputedShaderFlags &>
+  getShaderFlagsMask(const Function *) const;
+  const ComputedShaderFlags getCombinedFlags() const;
```
A few things here:
1. `getShaderFlagsMask` should just return a `const ComputedShaderFlags &`, not an expected. The only way for this to fail if the analysis is in a valid state is if we call this with a function that isn't in the module somehow.
2. `getCombinedFlags` should also return a reference - no need for a copy here.
3. These names are inconsistent (FlagsMask vs Flags) even though they return the same type of thing. Maybe we should simplify a little and call them `getFunctionFlags` and `getCombinedFlags`

bharadwajy replied:
Changes made per suggestions.

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -60,21 +62,54 @@ struct ComputedShaderFlags {
+  const ComputedShaderFlags getCombinedFlags() const;
+
+private:
+  /// Vector of Function-Shader Flag mask pairs representing properties of each
+  /// of the functions in the module. Shader Flags of each function represent
+  /// both module-level and function-level flags
```
Worth noting that this is a sorted vector.

bharadwajy replied:
> Worth noting that this is a sorted vector.

Updated comment accordingly.

bogner commented on llvm/lib/Target/DirectX/DXILShaderFlags.cpp:
```diff
@@ -13,36 +13,79 @@
+};
+} // namespace
+
+void DXILModuleShaderFlagsInfo::updateFunctionFlags(ComputedShaderFlags &CSF,
+                                                    const Instruction &I) {
```
Why is this a member function of `DXILModuleShaderFlagsInfo` and not just a static function? It doesn't use or change any state from the object.

bharadwajy replied:
> Why is this a member function of `DXILModuleShaderFlagsInfo` and not just a static function? It doesn't use or change any state from the object.

Changed it to a static function.

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:
```diff
@@ -1,27 +1,71 @@
+; CHECK-NEXT: ; Shader Flags mask for Function: test_uitofp_i64
+; CHECK-NEXT: ; Shader Flags Value: 0x00000044
+; CHECK-NEXT: ;
+; CHECK-NEXT: ; Note: shader requires additional functionality:
+; CHECK-NEXT: ;       Double-precision floating point
+; CHECK-NEXT: ;       Double-precision extensions for 11.1
+; CHECK-NEXT: ; Note: extra DXIL module flags:
```
It seems a bit ridiculous to print these notes after every function. Can't that be printed only for the module flags?

bharadwajy replied:
> It seems a bit ridiculous to print these notes after every function. Can't that be printed only for the module flags?

Changes made.

bogner reviewed (commented):
This is looking pretty close. As both Damyan and Chris pointed out, this isn't the appropriate kind of error handling here - the only way to hit the error is to break an invariant, which means we don't need user facing errors at all. A few other comments below.

bharadwajy commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -60,21 +62,54 @@ struct ComputedShaderFlags {
+  uint64_t getModuleFlags() const {
+    uint64_t ModuleFlags = 0;
+#define DXIL_MODULE_FLAG(DxilModuleBit, FlagName, Str)                         \
+  ModuleFlags |= FlagName ? getMask(DxilModuleBit) : 0ull;
```
> We have this exact expansion inside the `operator uint64_t` definition above. Should we maybe refactor this so that one uses the other instead of duplicating the expansion?

Refactored.

bharadwajy commented on llvm/lib/Target/DirectX/DXILShaderFlags.h:
```diff
@@ -60,21 +62,54 @@ struct ComputedShaderFlags {
   LLVM_DUMP_METHOD void dump() const { print(); }
 };
 
+struct DXILModuleShaderFlagsInfo {
```
> nit: This class's full name is `llvm::dxil::DXILModuleShaderFlagsInfo`, that's a bit of a mouthful. I'm not really sure how much benefit we get from prefixing it with DXIL and suffixing it with Info.

Changed to `ModuleShaderFlags`

Commit:
Rename getShaderFlagsMask() as getFunctionFlags()

Print Module flags notes only once.

Delete extraneous checks in test

Declare updateFunctionFlags as static

Use comparator instead of constructing a fuction-mask pair to
search in getFunctionFlags()

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions-obj-test.ll:
```diff
@@ -0,0 +1,17 @@
+; CHECK: - Name:            SFI0
+; CHECK-NEXT:     Size:            8
+; CHECK-NEXT:     Flags:
+; CHECK-NEXT:       Doubles:         true
+; CHECK-NOT:   {{[A-Za-z]+: +true}}
+; CHECK:            DX11_1_DoubleExtensions:         true
```
```suggestion
; CHECK: Doubles: true
; CHECK: DX11_1_DoubleExtensions: true
```

bharadwajy replied:
Changed.

bogner commented on llvm/lib/Target/DirectX/DXILTranslateMetadata.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "llvm/IR/Module.h"
 #include "llvm/InitializePasses.h"
 #include "llvm/Pass.h"
+#include "llvm/Support/Error.h"
```
Unused include

bharadwajy replied:
> Unused include

Deleted.

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions.ll:
```diff
@@ -1,27 +1,49 @@
-; CHECK-NEXT: {{^;$}}
-define double @div(double %a, double %b) #0 {
+; CHECK-NEXT: ;
+; CHECK-NEXT: ; Shader Flags for Module Functions
+; CHECK-NEXT: ; Function test_fdiv_double : 0x00000044
+; CHECK-NEXT: ;
+; CHECK-NEXT: ; Function test_uitofp_i64 : 0x00000044
+; CHECK-NEXT: ;
+; CHECK-NEXT: ; Function test_sitofp_i64 : 0x00000044
+; CHECK-NEXT: ;
+; CHECK-NEXT: ; Function test_fptoui_i32 : 0x00000044
+; CHECK-NEXT: ;
+; CHECK-NEXT: ; Function test_fptosi_i64 : 0x00000044
```
Might be easier to read/update this test if the checks for each function were by the function definition. I also think we can probably dispense of the `CHECK-NEXT` in that case and not bother checking the empty lines - if the checks are all by the function definitions it's obvious if we missed one.

bharadwajy replied:
> Might be easier to read/update this test if the checks for each function were by the function definition. I also think we can probably dispense of the `CHECK-NEXT` in that case and not bother checking the empty lines - if the checks are all by the function definitions it's obvious if we missed one.

Changes made.

Commit:
Cleanup of CHECK directives in tests.

bogner commented on llvm/test/CodeGen/DirectX/ShaderFlags/double-extensions-obj-test.ll:
```diff
@@ -11,7 +11,6 @@ attributes #0 = { convergent norecurse nounwind "hlsl.export"}
-; CHECK-NEXT:       Doubles:         true
-; CHECK-NOT:   {{[A-Za-z]+: +true}}
+; CHECK:       Doubles:         true
 ; CHECK:            DX11_1_DoubleExtensions:         true
```
It'd be nice to have consistent whitespace across the check lines

Commit:
Delete extraneous space in CHECK line
